### DISCUSSION TITLE
Help configure waveshare modbus CTS_602 hmi

### DISCUSSION THREAD

**Discussion by Martin21186** [veista/nilan#99] [Q&A] [answered]
Hello, could someone help me set up Modbus for Nilan CTS602 HMI? I have this Modbus Waveshare, but I don't know how to configure it correctly to send data. Thank you very much
![387792675_10222771528427070_3697693659730602644_n](https://github.com/veista/nilan/assets/113773943/9e385ec9-0615-438b-b30a-11b6f672c4e2)
![391687811_10222771528827080_1019854853108811183_n](https://github.com/veista/nilan/assets/113773943/68055270-0118-4334-9050-7fa4dc0e61f2)

**Comment by veista** [comment]
https://github.com/veista/nilan/discussions?discussions_q=waveshare

https://github.com/veista/nilan/issues?q=waveshare

**Comment by Martin21186** [comment]
I have a problem connecting in Homeassistant.  I don't know what to do anymore.  Do you think it's because I don't have modbus To eth from waveshare?  I am uploading my configuration.  If you can think of anything else, I will be glad.
![Screenshot_20231013_093455_Home Assistant](https://github.com/veista/nilan/assets/113773943/9c0c1bbe-288a-44c0-b870-c1cb11e848eb)
![Screenshot_20231013_093447_Home Assistant](https://github.com/veista/nilan/assets/113773943/83777d33-bdbc-481e-b020-8b4cba9eb27d)
![Screenshot_20231013_092129_Chrome](https://github.com/veista/nilan/assets/113773943/ecbb223e-aa6e-499b-8505-e3f12333b7cd)
![Screenshot_20231013_092123_Chrome](https://github.com/veista/nilan/assets/113773943/d47fb71a-8b6b-44da-9d7e-adbb7ee2b6de)

**Comment by veista** [comment]
> I have a problem connecting in Homeassistant. I don't know what to do anymore. Do you think it's because I don't have modbus To eth from waveshare? I am uploading my configuration. If you can think of anything else, I will be glad. 

I'm sorry, I don't understand. What don't you have?

**Reply by Martin21186** [reply]
I was wondering if a Waveshare rs485 to eth (b) modbus is needed?  I have an older modbus version.  Photo in previous text.  I connect to the modbus through the test program and it communicates and sends data from the PC.  Can't connect in Homeassistant.  Photo in previous text.  Where could be the problem?

**Reply by veista** [reply]
If you can communicate with for example qmodmaster it should be ok. You can debug with that.

**Reply by veista** [reply]
If that version can't translate modbus rtu to modbus tcp messages you need to edit the device.py file... change mode to rtuovertcp

**Reply by Martin21186** [reply]
Yes, when I sent the data for the test, on the main page you can see the IP address of the PC and the data sent.  PC connection works.  So how do you get modbus into homeassistant if you can't connect?

**Reply by veista** [reply]
Hi. There have been issues after 2023.9 version of HA.

The only solution is the keep trying if you are 100% sure everything should work. It works at random.

If you can connect with qmodmaster and get valid data and you have no other connections to the bridge. It only supports 1 connection. Just keep trying.

**Reply by veista** [reply]
> Yes, when I sent the data for the test, on the main page you can see the IP address of the PC and the data sent. 

Where can I see data?

**Reply by Martin21186** [reply]
So I downloaded the program and there it tells me a problem with the time.  The connection will go through, but the data probably cannot be sent.  Where can the problem be?  I don't know how to set the program exactly.  I did it according to youtube.  I'm an amateur at this.  I am very sorry but I would like to use your addon.  I will pay you for your time with a financial donation if it works out.  thank you very much
![20231013_204533](https://github.com/veista/nilan/assets/113773943/44654242-09c2-43a1-b5af-9f20f26878d3)

**Reply by Martin21186** [reply]
I am using the latest ha core 2023.10.2 and your addon 1.2.7.

**Reply by veista** [reply]
You need to get your bridge to translate modbus rtu messages to modbus tcp.

If your bridge can not do this you can try editing the device.py file and change from modbus tcp to rtuovertcp mode. This circumvents getting the proper equipment. I do not support it in the integration officially, but it is possible.

Timeout means it does not work. You can try editing the settings until qmodmaster says it works or you can try editing the file.

Just install vs code addon to HA for example and edit the file located in custom_components/nilan

**Reply by Martin21186** [reply]
Can you send a screenshot of where I can find the Modbus_TCP? I tried to look for it but couldn't find anything. I also tried to set up another Modbus in the QModMaster program that I have, and it's giving me some errors, so I probably have it configured incorrectly. Do you know how to do this? When you mentioned needing a bridge, what did you mean by that? Is there something else I need in addition to Modbus? Thank you.

**Reply by veista** [reply, marked as answer]
The waveshare device you have might not be a "bridge" it might only be a repeater.

A bridge translates modbus RTU messages (rs485 in this case) to Modbus TCP and vice versa. These are two different protocols which the device should bridge.

If you only have a repeater (I do not know) your waveshare device will send Modbus RTU messages over TCP. In this case you have to edit the code to support this.

Easier if you edit the config_flow.py file:

Replace line: self.data.update({"com_type": "tcp"})

With: self.data.update({"com_type": "rtuovertcp"})

This way you can receive updates without modifying the file every time.

**Reply by veista** [reply]
You also have to reboot HA after the edit.

**Reply by Martin21186** [reply]
I have this one.  I'll try it when I have some time.  Hope this helps.  Thank you
https://www.waveshare.com/rs485-to-eth.htm

**Reply by veista** [reply]
Yeah...

* This module does not support Modbus TCP, if you need to use this function, please purchase

**Reply by veista** [reply]
It wont work without the edit, it might not work at all.

**Reply by Martin21186** [reply]
Changing it didn't help either, so I'll go ahead and order a different Modbus converter. Is this one okay? Thank you.
https://www.waveshare.com/wiki/RS485_TO_POE_ETH_(B)

**Reply by veista** [reply]
I guess it should work, I do not know for sure.

**Reply by Martin21186** [reply]
After buying waveshare rs485 to eth (B) everything works as it should. Thanks for the information

**Comment by samupa** [comment]
Hello,

I have the same Waveshare Modbus repeater that sends Modbus RTU messages over TCP. I am able to read values from nilan with it, but Home assistant integration configuration fails even when I edit the config_flow.py file.

Here are my repeater settings and read values from nilan. 
![image](https://github.com/veista/nilan/assets/122469233/9757f318-9c42-4c5e-8f2b-b13e29feefc8)

Here is my edited config_flow.py file.
![image](https://github.com/veista/nilan/assets/122469233/f8448d0d-0bac-4fac-83cf-0c9358566dd7)

Here is failed integration.
![image](https://github.com/veista/nilan/assets/122469233/99aff08b-2bc8-44d5-b2bd-47f752836c5f)
![image](https://github.com/veista/nilan/assets/122469233/a9f5c1d1-1ac8-4e75-ad5e-8b25c463da33)

I am not really familiar with home assistant, but I didnt find anything from home-assistant.log or home-assistant.log.1
Is there some other log that would help?

**Reply by Martin21186** [reply]
It doesn't work with this device.  I bought a newer one and it worked the first time.  I also had to swap + and - for rs485.
https://www.waveshare.com/wiki/RS485_TO_POE_ETH_(B)

**Reply by samupa** [reply]
Thank you for reply. 
I would have bought that one if I would have read this thread fully.

I thought your problem was between nilan and waveshare? That you couldnt get any readings from modbus. I just cant with this integration. 

veista wrote this:
> If you only have a repeater (I do not know) your waveshare device will send Modbus RTU messages over TCP. In this case you have to edit the code to support this.
> 
> Easier if you edit the config_flow.py file:
> 
> Replace line: self.data.update({"com_type": "tcp"})
> 
> With: self.data.update({"com_type": "rtuovertcp"})
> 

I think its possible to use this device if integration supports RTU over TCP.

**Reply by veista** [reply]
The integration does not support it. I won't develop it, but technically it is possible. You will have to do all of the work by yourself.
